sdk/go/client.go

```diff
@@ -24,12 +25,56 @@ type GrpcClient struct {
 	conn *grpc.ClientConn
 }
 
+// ClientOption configures how we set up the grpc client
+type ClientOption interface {
+	apply(*clientOptions)
+}
+
+// WithTracer returns a ClientOption that set the tracer instance to be used by the grpc client.
+// If not specified, the tracer will be a no ops tracer by default.
+func WithTracer(tracer opentracing.Tracer) ClientOption {
+	return newFuncClientOption(func(o *clientOptions) {
+		o.tracer = tracer
+	})
+}
+
+type clientOptions struct {
+
+	tracer opentracing.Tracer
+}
+
+
+// funcClientOption wraps a function that modifies clientOptions into an
+// implementation of the ClientOption interface.
+type funcClientOption struct {
+	f func(*clientOptions)
+}
+
+func (fdo *funcClientOption) apply(do *clientOptions) {
+	fdo.f(do)
```

Comment:
Is all this complexity around ClientOptions just so that we can have an optional argument? Wouldn't it be cleaner to use some kind of getter/setter or builder pattern?